[FLINK-4248] [core] [table] CsvTableSource does not support reading SqlTimeTypeInfo types

twalthr opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR implements parsers for SQL Date/Time/Timestamp and tests. It also tests if the new parsers can be used in the RowCsvInputFormat.

gallenvara commented:
Hi @twalthr , the logical of parserFields method in Date/Time/TimeStamp (also in the lastest PR for BigDecimal/BigInteger) is same, is it better to refactor by creating super class for them to reduce the duplicate codes?

twalthr commented:
@gallenvara I thought the same. But I'm not sure if it is worth it. You could save ~25 lines of code per class maybe and if we want to make the parsing more efficient (do not use the `valueOf` methods, but own implementation) the super class becomes obsolete anyway.

fhueske commented on flink-core/src/test/java/org/apache/flink/types/parser/SqlDateParserTest.java:
```diff
@@ -0,0 +1,64 @@
+	@Override
+	public String[] getInvalidTestValues() {
+		return new String[] {
+			" 2013-08-12", "2013-08-12 ", "2013-08--12", "13-08-12", "2013/08/12", " ", "\t",
```
Add a value with an invalid date like `"2000-02-30"`?

twalthr replied:
The Java SQL date parser supports this and converts this into "2000-03-01".

fhueske replied:
What does it make out of `"2000-14-01"`?

twalthr replied:
That results in a `IllegalArgumentException`.

fhueske commented on flink-core/src/test/java/org/apache/flink/types/parser/SqlTimeParserTest.java:
```diff
@@ -0,0 +1,63 @@
+	@Override
+	public String[] getInvalidTestValues() {
+		return new String[] {
+			" 00:00:00", "00:00:00 ", "00:00::00", "00x00:00", "2013/08/12", " ", "\t"
```
Add an invalid time `"12:00:61"`?

twalthr replied:
This is also not invalid. The seconds are converted to minutes.

fhueske commented on flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java:
```diff
@@ -0,0 +1,69 @@
+	@Override
+	public String[] getValidTestValues() {
+		return new String[] {
+			"1970-01-01 00:00:00.000", "1990-10-14 02:42:25.123", "1990-10-14 02:42:25.123000001",
```
Add a valid time without second fractions: `"2016-05-01 12:34:50"`

fhueske commented on flink-core/src/test/java/org/apache/flink/types/parser/SqlTimestampParserTest.java:
```diff
@@ -0,0 +1,69 @@
+	public String[] getInvalidTestValues() {
+		return new String[] {
+			" 2013-08-12 14:15:59.479", "2013-08-12 14:15:59.479 ", "1970-01-01 00:00::00",
+			"00x00:00", "2013/08/12", "0000-01-01 00:00:00.f00", "2013-08-12 14:15:59.4788888888888888",
```
What is invalid about `"2013-08-12 14:15:59.4788888888888888"`, too much precision?
Add an invalid timestamp `"2010-12-12 15:67:01.123"`?

twalthr replied:
Yes, the maximum precision is 9 digits.
The timestamp you suggest is also valid.

fhueske commented on flink-core/src/main/java/org/apache/flink/types/parser/SqlDateParser.java:
```diff
@@ -0,0 +1,134 @@
+	public int parseField(byte[] bytes, int startPos, int limit, byte[] delimiter, Date reusable) {
+		int i = startPos;
+
+		final int delimLimit = limit - delimiter.length + 1;
```
Should we move the logic of creating a String without leading or tailing space to `FieldParser`? A couple of parsers use that (the ones in this PR, BigInteger, ...).

twalthr replied:
I will try to generalize it a bit.

twalthr commented:
@fhueske thanks for reviewing my code. I addressed your feedback.

fhueske commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
+
+		if (i > 0 &&
+				(Character.isWhitespace(bytes[startPos]) || Character.isWhitespace(bytes[startPos + i - 1]))) {
+			throw new NumberFormatException("There is leading or trailing whitespace in the numeric field.");
```
Throw a `IllegalArgumentException`? Method might be used by parsers that do not parse numeric values.

twalthr replied:
I think we could just rename the method again. Maybe `numericString()`?

fhueske replied:
But it does not do anything related to numeric values or am I overlooking something?

fhueske commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
+	 *
+	 * @return the parsed string
+	 */
+	public static final String formattedString(byte[] bytes, int startPos, int length, char delimiter) {
```
make `protected`?

fhueske replied:
What is formatted about the returned string? How about renaming the method to `nextFieldAsString()`?

fhueske commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
+	 *
+	 * @return the end position of the string or -1 if an error occurred
+	 */
+	public final int formattedStringEndPos(byte[] bytes, int startPos, int limit, byte[] delimiter) {
```
make `protected`?

fhueske replied:
rename to something like `findNextFieldEndPos()`?

fhueske commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
+	}
+
+	/**
+	 * Returns a string with a numeric format (like XXXX-XX-XX). Throws an exception if the
```
Why is this method specific for numeric formats?

twalthr replied:
Why not? Only parsers with a numeric format will use this method.

fhueske replied:
It returns a plain string object and only checks if the string has a leading or tailing whitespace. There are no checks for numeric characters or similiar. I think the method should be named according to what it does and not who's the intended caller.

fhueske commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
+	 */
+	public static final String formattedString(byte[] bytes, int startPos, int length, char delimiter) {
+		if (length <= 0) {
+			throw new NumberFormatException("Invalid input: Empty string");
```
Change to `IllegalArgumentException`?

twalthr replied:
According to the Javadoc "parseField" should return a `NumberFormatException`.

fhueske replied:
But this is not the `parseField` method. `parseField` could catch the exception and pass it on as a `NumberFormatException`.

fhueske commented on flink-core/src/main/java/org/apache/flink/types/parser/FieldParser.java:
```diff
@@ -174,6 +174,62 @@ protected void setErrorState(ParseErrorState error) {
 	}
+
+	/**
+	 * Returns the end position of a string with a numeric format (like XXXX-XX-XX). Sets the error state if the
```
Why numeric format?

fhueske replied:
Maybe reduce the scope of the method to only identify the end of the field and do the whitespace check in each parser separately?

twalthr commented:
@fhueske I think we should keep the methods as they currently are. I renamed them to "nextNumericString"/"nextNumericStringEndPos" and added a explanation what a numeric string is. You are right they don't do much numeric at the moment but only the numeric classes will use it. Furthermore, the error states and exceptions are numeric. E.g. if we don't do the whitespace checking in `nextNumericString` we also need to return the position instead of the string.

fhueske commented:
I still think that methods should be named after what they do (and there is nothing numeric about their behavior) and not in which context they are supposed to be called. The error messages are numeric due to the original context of the code which has been moved into separate methods. That context is no longer present in these methods. Anyway, I don't want to start a bikeshedding discussion about the names of internal utility methods and stop at this point ;-)
